Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch

```diff
@@ -0,0 +1,103 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: JRoy <joshroy126@gmail.com>
+Date: Wed, 26 Aug 2020 02:11:58 -0400
+Subject: [PATCH] Add additional open container api to HumanEntity
+
+
+diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
+index b09d12390d5f77330ac84452e0fee63a169bd01f..77bff8fb6bfdf739e413084e13677a83e723c71e 100644
+--- a/src/main/java/org/bukkit/entity/HumanEntity.java
++++ b/src/main/java/org/bukkit/entity/HumanEntity.java
+@@ -148,6 +148,92 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
+     @Nullable
+     public InventoryView openMerchant(@NotNull Merchant merchant, boolean force);
+ 
++    // Paper start - Add additional containers
++    /**
++     * Opens an empty anvil inventory window with the player's inventory
++     * on the bottom.
++     *
++     * @param location The location to attach it to. If null, the player's
++     *     location is used.
++     * @param force If false, and there is no anvil block at the location,
++     *     no inventory will be opened and null will be returned.
++     * @return The newly opened inventory view, or null if it could not be
++     *     opened.
++     */
++    @Nullable
++    public InventoryView openAnvil(@Nullable Location location, boolean force);
++
++    /**
++     * Opens an empty cartography table inventory window with the player's inventory
++     * on the bottom.
++     *
++     * @param location The location to attach it to. If null, the player's
++     *     location is used.
++     * @param force If false, and there is no cartography table block at the location,
++     *     no inventory will be opened and null will be returned.
++     * @return The newly opened inventory view, or null if it could not be
++     *     opened.
++     */
++    @Nullable
++    public InventoryView openCartographyTable(@Nullable Location location, boolean force);
```

Comment:
Wouldn't it be easier for users to just specify inventory type?

Reply:
Existing API to do this doesn't work correctly for these use cases (UnsupportedOperationExceptions + improper reachability checks). Due to the way containers are written, speific casting is required in order to properly so one general purpose method would end up being as long as indivial methods.